File: 45-jump-game-ii/45-jump-game-ii.cpp
class Solution {
public:
    vector<int> nums;
    int dp[10007];
    int solveIt(int idx) {
        if (idx >= nums.size() - 1)
            return 0;
        
        if(dp[idx] != -1) return dp[idx];
        
        int res = 1000000007;
        for (int jump = 1; jump <= nums[idx]; jump++) {
            res = min(res, 1 + solveIt(idx + jump));
        }
        
        return dp[idx] = res;
    }
    int jump(vector<int>& nums) {
        this -> nums = nums;
        memset(dp, -1, sizeof dp);
        return solveIt(0);
    }
};

File: 45-jump-game-ii/45-jump-game-ii.java
class Solution {
    int n;
    int[] dp = new int[10004];
    int solve(int idx, int[] nums) {
        if (idx == n - 1)
            return 0;
        
        if (dp[idx] != -1) return dp[idx];
        
        int maxJump = nums[idx];
        
        int res = 99999999;
        for (int i = 1 ; i <= maxJump; i++) {
            int nextIdx = i + idx;
            if (nextIdx >= n)
                break;
            
            int option =  1 + solve(nextIdx, nums);
            res = Math.min(res, option);    
        }
        
        dp[idx] = res;
        return res;
    }
    public int jump(int[] nums) {
        n = nums.length;
        
        Arrays.fill(dp, -1);
        return solve(0, nums);
    }
}